lib/public_key/test/public_key_SUITE.erl

```diff
@@ -926,7 +929,47 @@ pkix_path_validation_root_expired(Config) when is_list(Config) ->
     true = public_key:pkix_is_self_signed(Root),
     Peer = proplists:get_value(cert, Conf),
     {error, {bad_cert, cert_expired}} = public_key:pkix_path_validation(Root, [ICA, Peer], []).
+
+
+pkix_path_validation_bad_date() ->
+    [{doc, "Ensure bad date formats in `validity` are handled gracefully by verify fun"}].
+pkix_path_validation_bad_date(Config) when is_list(Config) ->
+    
+    % Load PEM certchain from file
+    DataDir = proplists:get_value(data_dir, Config),
+    {ok, Bin} = file:read_file(filename:join(DataDir,"bad_date_certchain.pem")),
+
+    % Decode and extract raw der encoded certificates
+    CertificateList = public_key:pem_decode(Bin),
+    [Root | CertificateChain] = lists:map(fun({'Certificate', Der, _}) -> Der end, CertificateList),
     
+    % First test exception throwing without validation function
+    try
+        public_key:pkix_path_validation(Root, CertificateChain, [])
+    catch
+        % We expect this function to throw if the date is wrong
+        error:function_clause ->
+            ok;
+        _ ->
+            throw("Expected function_clause error on bad date")
+    end,
```

Comment:
This is the current behaviour if no verify fun is supplied. Why would you like to change it?

Reply:
Since the test will be merged along with your changes there is no reason to test the behavior in the old code.

You shall test the new functionality, i.e. that {error, {bad_cert, invalid_validity_dates}} is returned if no verify_fun is provided.

Reply:
@andreasanta what about this one?

Reply:
@dgud This diff does not alter the old functionality if you don't provide `verify_fun`. This is why I am leaving the testcase. In order to preserve backwards compatibility, I am still raising an error if no `verify_fun` is specified.

@u3s Sorry maybe I did not explain my diff correctly. The previous behaviour has not been altered on purpose. You can optionally alter if if you pass a `verify_fun` that intercepts the validation error.  This testcase exists to ensure that the previous behaviour - i.e. throwing an error upon failed validation - still happens if you don't specify a `verify_fun`.

Please let me know if you'd like a different approach.

Reply:
You have altered the functionality if no verify_fun is used and you fail to test it.

Reply:
I do not see a reason to preserve the crash behavior for something that should not really happen.  I think this PR is about  adding a workaround | graceful handling of that it sometimes does due to mistakes that are out of users control.

Reply:
@IngelaAndin Refactored as suggested, plus some other improvements, to improve readability.

Reply:
> You have altered the functionality if no verify_fun is used and you fail to test it.

Sorry I do not understand it. I have written two tests, one that lets the code crash as expected, the other prevents the crash by supplying a custom `verify_fun`. What did I miss?